File: mergeSort.cpp
#include <iostream>
using namespace std;
void merge(int arr[], int low, int mid, int high)
{
  int i = low, j = mid + 1, k = low, z[20];
  while (i <= mid && j <= high)
  {
    if (arr[i] <= arr[j])
      z[k++] = arr[i++];
    else
      z[k++] = arr[j++];
  }
  while (i <= mid)
    z[k++] = arr[i++];
  while (j <= high)
    z[k++] = arr[j++];
  for (i = low; i <= high; i++)
    arr[i] = z[i];
}
void mergeSort(int *arr, int low, int high)
{
  if (low < high)
  {
    int mid = (low + high) / 2;
    mergeSort(arr, low, mid);

    mergeSort(arr, mid + 1, high);
    merge(arr, low, mid, high);
  }
}
int main()
{
  int n;
  cin >> n;
  int *arr = new int[n];
  for (int i = 0; i < n; i++)
  {
    cin >> arr[i];
  }
  mergeSort(arr, 0, n - 1);
  cout << "Sorted array is" << endl;
  for (int i = 0; i < n; i++)
    cout << arr[i] << " ";

  return 0;
}

File: Basic DSA/MergeSort.java
public class MergeSort {
	
	public static void _sort(int[] arr) {
		int beg = 0;
		int end = arr.length-1;
		mergeSort(arr, beg, end);
	}
    
    private static void mergeSort(int[] arr, int beg, int end) {
        
        if(beg<end){
            int mid = (beg+end)/2;
            mergeSort(arr, beg, mid);
            mergeSort(arr, mid+1, end);
            merge(arr, beg, mid, end);
        }
    }

    private static void merge(int[] arr, int beg, int mid, int end) {

        int n1= mid-beg+1;
        int n2= end-mid;

        int L[] = new int[n1];
        int R[] = new int[n2];

        for(int i=0; i<n1; i++)
            L[i]=arr[beg+i];
        for(int j=0;j<n2;j++)
            R[j]=arr[mid+1+j];

        int i=0, j=0;
        int k= beg;

        while(i<n1 && j<n2) {

            if(L[i]<=R[j]) {
                arr[k]=L[i];
                i++;
            }else {
                arr[k]=R[j];
                j++;
            }
            k++;
        }

        while(i<n1) {
            arr[k]=L[i];
            i++;
            k++;
        }

        while(j<n2) {
            arr[k]=R[j];
            j++;
            k++;
        }
    }
}